# Week 8 | Project - 3
Group 4: Joshua Hummell, Jiho Kim, Scott Reed



https://www.youtube.com/watch?v=75mxuo8w-UE

Using any of the three classifiers described in chapter 6 of Natural Language Processing with Python,
and any features you can think of, build the best name gender classifier you can.

In [7]:
import importlib.util
import random
from nltk.classify.naivebayes import NaiveBayesClassifier
import nltk.classify

spec = importlib.util.find_spec('nltk')
if spec is None:
    print(package_name +" is not installed. Installing:")
    nltk.download()
else:
    (print("Exists."))
from nltk.corpus import names

Exists.


In [8]:
names = ([(name, 'male') for name in names.words('male.txt')] +
          [(name, 'female') for name in names.words('female.txt')])

random.Random(512).shuffle(names)

In [37]:
print(names[:5])

[('Dolorita', 'female'), ('Kennedy', 'male'), ('Nessi', 'female'), ('Karie', 'female'), ('Florie', 'female')]


In [35]:
def tryFeatures(f, c = NaiveBayesClassifier, final= False):
     #Begin by splitting the Names Corpus into three subsets: 500 words for the test set, 
     #500 words for the devtest set, and the remaining 6900 words for the training set.
    test_names, devtest_names, train_names = names[:499], names[500:999], names[1000:]
    
    train_set = [(f(n), g) for (n,g) in train_names]
    
    devtest_set = [(f(n), g) for (n,g) in devtest_names]
    
    test_set = [(f(n), g) for (n,g) in test_names]
    
    classifier = c.train(train_set) 
    
    testSet = devtest_set
    
    if(final):
        testSet = test_set
        
    return (nltk.classify.accuracy(classifier, train_set),nltk.classify.accuracy(classifier, testSet)) 


In [36]:
print(tryFeatures(lambda a: {'len':len(a), 'last':a[-2:], 'trim':a[-1:], 'start':a[:3]}))

(0.857286866359447, 0.8156312625250501)


In [25]:
tryFeatures(lambda a: {'len':len(a), 'last':a[-2:], 'trim':a[-1:], 'start':a[:3]},nltk.classify.DecisionTreeClassifier)

(0.955213133640553, 0.736)

Then, starting with the example name gender classifier, make incremental improvements. Use the dev-test set to check your progress.

In [26]:
tryFeatures(lambda a: {'len':len(a), 'last':a[-2:], 'trim':a[-1:], 'start':a[:3]},nltk.classify.MaxentClassifier)

  ==> Training (100 iterations)

      Iteration    Log Likelihood    Accuracy
      ---------------------------------------
             1          -0.69315        0.372
             2          -0.38613        0.818
             3          -0.32632        0.844
             4          -0.29786        0.857
             5          -0.28027        0.862
             6          -0.26781        0.868
             7          -0.25825        0.871
             8          -0.25054        0.875
             9          -0.24412        0.879
            10          -0.23864        0.882
            11          -0.23389        0.885
            12          -0.22972        0.886
            13          -0.22600        0.887
            14          -0.22267        0.889
            15          -0.21967        0.890
            16          -0.21693        0.891
            17          -0.21443        0.892
            18          -0.21213        0.892
            19          -0.21000        0.893
 

(0.9004896313364056, 0.828)

Once you are
satisfied with your classifier, check its final performance on the test set.
How does the performance on the test set compare to the performance on the dev-test set? Is this what
you'd expect?

Looking at the results, a descision tree classifier over fits the data getting 95% accuracy on the training set but only 73.6% on the test set, while a naive bayes classifier does surprisingly well, attaining 90% accuracy on the 100th try. This suggests we use it for the final check. 

In [18]:
print(tryFeatures(lambda a: {'len':len(a), 'last':a[-2:], 'trim':a[-1:], 'start':a[:3]}, final =True))

(0.8609559279950342, 0.838)


This does quite well, actually better than devtest (83.8% vs 81.6%). Normally one would expect a regression. 